File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/ServiceTypesHelper.java
package ru.tinkoff.kora.kora.app.annotation.processor;

import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;
import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;
import ru.tinkoff.kora.annotation.processor.common.NameUtils;

import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.Objects;

public class ServiceTypesHelper {
    private final Elements elements;
    private final Types types;
    private final TypeElement wrappedTypeElement;
    private final DeclaredType wrappedType;
    private final TypeElement interceptorTypeElement;
    private final DeclaredType interceptorType;

    public ServiceTypesHelper(Elements elements, Types types) {
        this.elements = elements;
        this.types = types;
        this.wrappedTypeElement = Objects.requireNonNull(this.elements.getTypeElement(CommonClassNames.wrapped.canonicalName()));
        this.wrappedType = Objects.requireNonNull(this.types.getDeclaredType(this.wrappedTypeElement, this.types.getWildcardType(null, null)));
        this.interceptorTypeElement = Objects.requireNonNull(this.elements.getTypeElement(CommonClassNames.graphInterceptor.canonicalName()));
        this.interceptorType = Objects.requireNonNull(this.types.getDeclaredType(this.interceptorTypeElement, this.types.getWildcardType(null, null)));
    }

    public boolean isAssignableToUnwrapped(TypeMirror maybeWrapped, TypeMirror typeMirror) {
        if (!this.types.isAssignable(maybeWrapped, this.wrappedType)) {
            return false;
        }
        var wrappedParameterElement = this.wrappedTypeElement.getTypeParameters().get(0); // somehow it can be changed during execution
        var declaredType = (DeclaredType) maybeWrapped;
        var unwrappedType = this.types.asMemberOf(declaredType, wrappedParameterElement);
        return this.types.isAssignable(unwrappedType, typeMirror);
    }

    public boolean isSameToUnwrapped(TypeMirror maybeWrapped, TypeMirror typeMirror) {
        if (!this.types.isAssignable(maybeWrapped, this.wrappedType)) {
            return false;
        }
        var wrappedParameterElement = this.wrappedTypeElement.getTypeParameters().get(0); // somehow it can be changed during execution
        var declaredType = (DeclaredType) maybeWrapped;
        var unwrappedType = this.types.asMemberOf(declaredType, wrappedParameterElement);
        return this.types.isSameType(unwrappedType, typeMirror);
    }

    public boolean isInterceptorFor(TypeMirror interceptorType, TypeMirror targetType) {
        if (!this.types.isAssignable(interceptorType, this.interceptorType)) {
            return false;
        }
        var interceptorTypeParameter = this.interceptorTypeElement.getTypeParameters().get(0); // somehow it can be changed during execution
        var declaredType = (DeclaredType) interceptorType;
        var interceptedType = this.types.asMemberOf(declaredType, interceptorTypeParameter);
        return isInterceptable(interceptedType, targetType);
    }

    public boolean isInterceptable(TypeMirror interceptedType, TypeMirror targetType) {
        if (this.types.isSameType(interceptedType, targetType)) {
            return true;
        } else if (this.types.isAssignable(targetType, interceptedType)) {
            // Check if is AopProxy
            var typeMirrorElement = types.asElement(targetType);
            var annotation = AnnotationUtils.findAnnotation(typeMirrorElement, CommonClassNames.aopProxy);
            if (annotation != null) {
                var interceptedTypeElement = types.asElement(interceptedType);
                var aopProxyName = NameUtils.generatedType(interceptedTypeElement, "_AopProxy");
                var expectedAopProxyCanonicalName = elements.getPackageOf(interceptedTypeElement).toString() + "." + aopProxyName;
                return expectedAopProxyCanonicalName.equals(typeMirrorElement.toString());
            }
        }

        return false;
    }

    public boolean isInterceptor(TypeMirror maybeInterceptor) {
        return this.types.isAssignable(maybeInterceptor, this.interceptorType);
    }

    public TypeMirror getInterceptedType(TypeMirror maybeInterceptor) {
        var interceptorTypeParameter = this.interceptorTypeElement.getTypeParameters().get(0); // somehow it can be changed during execution
        var declaredType = (DeclaredType) maybeInterceptor;
        return this.types.asMemberOf(declaredType, interceptorTypeParameter);
    }
}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/ServiceTypesHelper.kt
package ru.tinkoff.kora.kora.app.ksp

import com.google.devtools.ksp.getAllSuperTypes
import com.google.devtools.ksp.getDeclaredFunctions
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.ClassKind
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.CommonClassNames.isVoid
import ru.tinkoff.kora.ksp.common.getOuterClassesAsPrefix


class ServiceTypesHelper(val resolver: Resolver) {
    private val interceptorClassDeclaration = resolver.getClassDeclarationByName(resolver.getKSNameFromString(CommonClassNames.graphInterceptor.canonicalName))!!
    private val interceptorType = interceptorClassDeclaration.asStarProjectedType()
    private val interceptorInitFunction = interceptorClassDeclaration.getDeclaredFunctions()
        .filter { it.simpleName.asString() == "init" && it.parameters.size == 1 }
        .first()


    private val wrappedClassDeclaration = resolver.getClassDeclarationByName(resolver.getKSNameFromString(CommonClassNames.wrapped.canonicalName))!!
    private val wrappedType = wrappedClassDeclaration.asStarProjectedType()
    private val wrappedValueFunction = wrappedClassDeclaration.getDeclaredFunctions()
        .filter { it.simpleName.asString() == "value" && it.parameters.isEmpty() }
        .first()

    companion object {

        fun findAopProxySuperClass(maybeAopProxy: KSType): KSType? {
            val aopProxyAnnotation = maybeAopProxy.declaration.findAnnotation(CommonClassNames.aopProxy)
            val proxyDeclaration = maybeAopProxy.declaration
            if (aopProxyAnnotation != null && proxyDeclaration is KSClassDeclaration) {
                val proxyParent = (maybeAopProxy.declaration as KSClassDeclaration).superTypes
                    .map { it.resolve() }
                    .filter { (it.declaration as? KSClassDeclaration)?.classKind == ClassKind.CLASS }
                    .firstOrNull()

                if (proxyParent != null) {
                    val proxyParentDeclaration = proxyParent.declaration
                    val aopProxyName = proxyParentDeclaration.getOuterClassesAsPrefix() + proxyParentDeclaration.simpleName.asString() + "__AopProxy"
                    val aopProxyCanonical = proxyParentDeclaration.packageName.asString() + "." + aopProxyName
                    if (aopProxyCanonical == maybeAopProxy.declaration.qualifiedName!!.asString()) {
                        return proxyParent
                    }
                }
            }

            return null
        }
    }

    fun isAssignableToUnwrapped(maybeWrapped: KSType, type: KSType): Boolean {
        if (!wrappedType.isAssignableFrom(maybeWrapped)) {
            return false
        }
        val maybeWrappedDeclaration = maybeWrapped.declaration as KSClassDeclaration
        val wrappedClassDeclaration = maybeWrappedDeclaration.getAllSuperTypes().plus(sequence { this.yield(maybeWrappedDeclaration.asType(listOf())) })
            .first { CommonClassNames.wrapped.canonicalName == it.declaration.qualifiedName?.asString() }
            .declaration as KSClassDeclaration
        val wrappedValueFunction = wrappedClassDeclaration.getAllFunctions()
            .filter { it.simpleName.asString() == "value" }
            .first()
        val unwrappedType = wrappedValueFunction.asMemberOf(maybeWrapped).returnType!!
        return type.isAssignableFrom(unwrappedType)
    }

    fun isSameToUnwrapped(maybeWrapped: KSType, type: KSType): Boolean {
        if (!wrappedType.isAssignableFrom(maybeWrapped)) {
            return false
        }
        //TODO check if also is not working?
        val unwrappedType = wrappedValueFunction.asMemberOf(maybeWrapped).returnType!!
        return unwrappedType.makeNotNullable() == type // platform nullability ruins equality
    }

    fun isInterceptorFor(maybeInterceptor: KSType, type: KSType): Boolean {
        if (!interceptorType.isAssignableFrom(maybeInterceptor)) {
            return false
        }

        return try {
            val interceptType = interceptType(maybeInterceptor)
            return isInterceptable(interceptType, type)
        } catch (e: IllegalArgumentException) {
            false
        }
    }

    fun isInterceptable(interceptedType: KSType, targetType: KSType): Boolean {
        if (targetType == interceptedType.makeNotNullable()) {
            return true
        } else if (interceptedType.isAssignableFrom(targetType)) {
            val aopProxyAnnotation = targetType.declaration.findAnnotation(CommonClassNames.aopProxy)
            val proxyDeclaration = interceptedType.declaration
            if (aopProxyAnnotation != null && proxyDeclaration.parentDeclaration != null) {
                val aopProxyName = proxyDeclaration.getOuterClassesAsPrefix() + proxyDeclaration.simpleName.asString() + "__AopProxy"
                val aopProxyCanonical = proxyDeclaration.packageName.asString() + "." + aopProxyName
                return aopProxyCanonical == targetType.declaration.qualifiedName!!.asString()
            }
        }

        return false
    }

    fun interceptType(maybeInterceptor: KSType): KSType {
        if (!interceptorType.isAssignableFrom(maybeInterceptor)) {
            throw IllegalArgumentException()
        }

        return if (maybeInterceptor.declaration is KSClassDeclaration) {
            (maybeInterceptor.declaration as KSClassDeclaration).getDeclaredFunctions()
                .filter { f -> f.simpleName.asString() == "init" && f.parameters.size == 1 && f.returnType != null && !f.returnType!!.isVoid() }
                .filter { f -> f.parameters.first().type.toTypeName() == f.returnType!!.toTypeName() }
                .map { it.returnType!!.resolve() }
                .firstOrNull() ?: throw IllegalArgumentException()
        } else {
            val memberOf = interceptorInitFunction.asMemberOf(maybeInterceptor)
            return memberOf.parameterTypes[0]!!.makeNotNullable()
        }
    }

    fun isInterceptor(type: KSType) = interceptorType.isAssignableFrom(type)
}
